man/marked-man.1
.P
NPM \fIhttps://nodei\.co/npm/marked\-man\.png?downloads=true\fR \fIhttps://nodei\.co/npm/marked\-man/\fR
.TH "MARKED\-MAN" "1" "June 2019" "" ""
.SH "NAME"
\fBmarked-man\fR \- markdown to roff
.SH SYNOPSIS
.P
.RS 2
.nf
marked\-man README\.md > doc/marked\-man\.1
.fi
.RE
.P
See marked README \fIhttps://github\.com/chjj/marked\fR for documentation about how to use marked\.
.P
Note that \fBmarked\-man \-\-format=html\fP is the same as \fBmarked\fP\|\.
.SH DESCRIPTION
.P
\fBmarked\-man\fP wraps \fBmarked\fP to extend it with groff output support in order to
create Unix manual pages for use with \fBman\fP\|\.
.SH OPTIONS
.P
\fBmarked\-man\fP invokes \fBmarked \-\-gfm \-\-sanitize\fP, and you can pass additional
options through\.
.br
The \fB\-\-breaks\fP option, which retains intra\-paragraph line breaks, can be helpful to match default ronn behavior\.
.P
\fBmarked\-man\fP adds some options to \fBmarked\fP\|'s existing options:
.RS 0
.IP \(bu 2
\fB\-\-format <format>\fP
.br
Sets the output format\. Outputs html if different from \fBroff\fP\|\.  
Defaults to \fBroff\fP\|\.
.IP \(bu 2
\fB\-\-name <name>\fP
.br
The name shown in the manpage header, if it isn't given in the ronn header like in this README\.
.br
Defaults to empty string\.
.IP \(bu 2
\fB\-\-section <section>\fP
.br
The section number shown in the manpage header, if it isn't given in the ronn header like in this README\.
.br
Defaults to empty string\.
.IP \(bu 2
\fB\-\-ver <version>\fP
.br
The version shown in the manpage footer\.
.br
Defaults to empty string\.
.IP \(bu 2
\fB\-\-manual <manual>\fP
.br
The manual\-group name shown in the manpage header\.
.br
Defaults to empty string\.
.IP \(bu 2
\fB\-\-date <date>\fP
.br
The date shown in the manpage header\.
.br
Defaults to now, must be acceptable to \fBnew Date(string or timestamp)\fP\|\.

.RE
.SH INSTALLATION
.P
From the npm registry \fIhttps://npmjs\.com\fR:
.RS 0
.IP \(bu 2
locally (\fB\-\-save\fP, \fB\-\-save\-dev\fP, or \fB\-\-save\-optional\fP add \fBmarked\-man\fP to your \fBpackage\.json\fP file as a runtime, development\-time, or optional runtime dependency, respectively)
.P
.RS 2
.nf
  npm install marked\-man [\-\-save|\-\-save\-dev|\-\-save\-optional]
.fi
.RE
.IP \(bu 2
globally (puts \fBmarked\-man\fP in your system's path):
.P
.RS 2
.nf
  [sudo] npm install marked\-man \-g
.fi
.RE

.RE
.SH EXAMPLE
.P
To view this README as a man page, run something like the following:
.P
.RS 2
.nf
marked\-man \-\-ver v0\.1\.0 \-\-manual 'Man Utilities' README\.md > /tmp/mm$$\.1 && man /tmp/mm$$\.1
.fi
.RE
.SH SEE ALSO
.P
Ronn \fIhttps://github\.com/rtomayko/ronn\fR
.SH REPORTING BUGS
.P
See marked\-man repository \fIhttps://github\.com/kapouer/marked\-man\fR\|\.


CHANGES.md
0.1.4
=====

* depends on marked 0.3.2
* use .nf/.fi instead of .EX/.EE to format blockquotes

0.1.6
=====

* ensure tests pass

0.2.0
=====

* add support for gfm table

0.2.1
=====

* remove useless test files from npm package

0.3.0
=====

* accept a timestamp in milliseconds as --date argument

0.7.0
=====

* renamed `--version` to `--ver` to avoid collision with `marked`



lib/marked-man.js

const marked = require("marked"),
	InlineLexer = marked.InlineLexer,
	Parser = marked.Parser;

InlineLexer.prototype.html_output = InlineLexer.prototype.output;
InlineLexer.prototype.html_outputLink = InlineLexer.prototype.outputLink;

InlineLexer.prototype.roff_outputLink = function roff_outputLink(cap, link) {
	var href = resc(link.href),
		same = link.title == href,
		text = '';

	if (same) {
		text = '';
	} else if (!link.title) {
		text = this.output(cap[1]);
	}

	if (text) text += ' ';

	return text + "\\fI" + href + "\\fR";
};

InlineLexer.prototype.roff_output = function roff_output(src) {
	var out = '',
		link,
		text,
		href, //OBSOLETE(?)
		cap,
		cur;

	while (src) {
		// escape
		if (cap = this.rules.escape.exec(src)) {
			src = src.substring(cap[0].length);
			out += lastMatch(cap);
			continue;
		}

		// autolink
		if (cap = this.rules.autolink.exec(src)) {
			src = src.substring(cap[0].length);

			if (cap[2] === '@') {
				text = cap[1][6] === ':' ? cap[1].substring(7) : cap[1];
			} else {
				text = cap[1];
			}

			out += resc(text);
			continue;
		}

		// url (gfm)
		if (cap = this.rules.url.exec(src)) {
			src = src.substring(cap[0].length);
			out += resc(cap[0]);
			continue;
		}

		// tag
		if (cap = this.rules.tag.exec(src)) {
			src = src.substring(cap[0].length);
			out += resc(cap[0]);
			continue;
		}

		// link
		if (cap = this.rules.link.exec(src)) {
			src = src.substring(cap[0].length);
			out += this.outputLink(cap, {
				href: cap[2],
				title: cap[3]
			});
			continue;
		}

		// reflink, nolink
		if ((cap = this.rules.reflink.exec(src)) ||
			(cap = this.rules.nolink.exec(src))) {
			src = src.substring(cap[0].length);
			cur = lastMatch(cap).replace(/\s+/g, ' ');
			link = this.links[cur.toLowerCase()];

			if (!link || !link.href) {
				out += cap[0][0];
				src = cap[0].substring(1) + src;
				continue;
			} else if (link.title) {
				out = out.slice(0, -cur.length);
			}

			out += this.outputLink(cap, link);
			continue;
		}

		// strong
		if (cap = this.rules.strong.exec(src)) {
			src = src.substring(cap[0].length);
			out += '\\fB' + this.output(lastMatch(cap)) + '\\fR';
			continue;
		}

		// em
		if (cap = this.rules.em.exec(src)) {
			src = src.substring(cap[0].length);
			out += '\\fI' + this.output(lastMatch(cap)) + '\\fR';
			continue;
		}

		// code
		if (cap = this.rules.code.exec(src)) {
			src = src.substring(cap[0].length);
			out += '\\fB' + resc(lastMatch(cap), true) + '\\fP';
			continue;
		}

		// br
		if (cap = this.rules.br.exec(src)) {
			src = src.substring(cap[0].length);
			out += '\n.br\n';
			continue;
		}

		// del (gfm)
		if (cap = this.rules.del.exec(src)) {
			src = src.substring(cap[0].length);
			out += "-" + resc(lastMatch(cap)) + "-";
			continue;
		}

		// text
		if (cap = this.rules.text.exec(src)) {
			src = src.substring(cap[0].length);
			out += resc(cap[0]);
			continue;
		}

		if (src) {
			throw new Error('Infinite loop on byte: ' + src.charCodeAt(0));
		}
	}

	return out;
};

Parser.prototype.parse = function parse(src) {
	this.inline = new InlineLexer(src.links, this.options);
	this.tokens = src.reverse();

	if (this.options.format == "roff") {
		this.tok = this.roff_tok;
		this.inline.output = this.inline.roff_output;
		this.inline.outputLink = this.inline.roff_outputLink;

		var first = this.peek();

		if ((first.type != "heading" || first.depth != 1) && this.options.name) this.tokens.push({
			type: "heading",
			depth: 1
		});
	} else {
		this.tok = this.html_tok;
		this.inline.output = this.inline.html_output;
		this.inline.outputLink = this.inline.html_outputLink;
	}

	// NOTE marked.js Parser.tok() relies on a this.inlineText reference.
	this.inlineText = this.inline;

	var out = '';

	while (this.next()) {
		out += this.tok();
	}

	return out;
};

Parser.prototype.html_tok = Parser.prototype.tok;

Parser.prototype.isNestList = false;

Parser.prototype.roff_tok = function () {
	switch (this.token.type) {
		case "space": {
			return '';
		}
		case "hr": {
			return ".HR\n";
		}
		case "heading": {
			var macro, text = this.token.text;

			if (this.token.depth == 1) {
				macro = "TH";
				text = rparseHeader(text || "", this.options);
			} else if (this.token.depth == 2) {
				macro = "SH";
				text = this.inline.output(text);
			} else {
				macro = "SS";
				text = this.inline.output(text);
			}
			return '.' + macro + ' ' + text + '\n';
		}
		case "code": {
			if (this.options.highlight) {
				var code = this.options.highlight(this.token.text, this.token.lang);

				if (code != null && code !== this.token.text) {
					this.token.escaped = true;
					this.token.text = code;
				}
			}

			if (!this.token.escaped) {
				this.token.text = resc(this.token.text);
			}

			return ".P\n" + ".RS 2\n" + ".nf\n" + this.token.text + "\n.fi" + "\n.RE\n";
		}
		case "blockquote_start": {
			var body = '';

			while (this.next().type !== "blockquote_end") {
				body += this.tok();
			}

			return ".QP\n" +
				body +
				"\n.\n";
		}
		case "list_start": {
			var body = '';
			var order = this.token.ordered ? 1 : null;
			var originalLevel = this.isNestList;

			this.isNestList = true;

			while (this.next().type !== "list_end") {
				if (order) this.token.order = order++;
				body += this.tok();
			}

			this.isNestList = originalLevel;

			var indent = this.isNestList ? ".RS\n" : ".RS 0\n";

			return indent + body + "\n.RE\n";
		}
		case "loose_item_start":
		case "list_item_start": {
			var body = '',
				bullet = "\\(bu",
				offset = 2;
			
			if (this.token.order) {
				bullet = this.token.order + '.';
				offset = 3;
			}

			while (this.next().type !== "list_item_end") {
				var tok = this.tok();

				if (this.options.ronn && !this.options.breaks) {
					// replace first line that ends with colons
					tok = tok.replace(":\n", "\n.br\n");
					// replace dot + newline | whitespaces with dot + line break
					tok = tok.replace(/\.(?:\n|\s\s+)/g, ".\n.br\n");
				}

				body += tok;
			}

			return ".IP " + bullet + ' ' + offset + '\n' + body;
		}
		case "html": {
			return !this.token.pre && !this.options.pedantic ? this.inline.output(this.token.text) : this.token.text;
		}
		case "paragraph": {
			return ".P\n" + this.inline.output(this.token.text) + '\n';
		}
		case "text": {
			return '' + this.parseText() + '\n';
		}

		/* deal with table format, add by gholk.
		 * use **tbl** macro format table in man page.
		 */
		case "table": {
			var tblText = [ ], // troff tbl macro text.
				rowArray = [ ]; // tmp array store cells in a table row.

			tblText.push(".TS"); // table start

			tblText.push("tab(|) expand nowarn box;");
			/* set `|` as seperator.
			 *
			 * set table border.
			 *
			 * set expand to full screen. without expand,
			 * tbl sometimes mistake terminal width.
			 * but expand is ugly sometimes...
			 *
			 * command need end by `;`.
			 */

			var columnNumberRow = '';

			for (var i = 0; i < this.token.header.length; i++) {
				columnNumberRow += ' l';
				rowArray.push("T{\n" + this.inline.output(this.token.header[i]) + "\nT}");
			}

			columnNumberRow += '.';
			tblText.push(columnNumberRow);
			tblText.push(rowArray.join('|'));

			tblText.push('_'); // add an horizantle line after header


			/* run each row */
			for (var i = 0; i < this.token.cells.length; i++) {
				var thisRow = this.token.cells[i]; // pointer to old array.

				rowArray = [ ]; // new array store cells in a row.

				/* run each cell */
				for (var j = 0; j < thisRow.length; j++)
					rowArray.push("T{\n" + this.inline.output(thisRow[j]) + "\nT}");

				tblText.push(rowArray.join('|'));
			}

			tblText.push(".TE");

			return tblText.join('\n') + '\n';
		}
	}
};

function lastMatch(cap, exclude) {
	var len = cap.length,
		cur;
	
	while (--len >= 0) {
		cur = cap[len];
		if (cur === exclude) continue;
		if (cur !== undefined) return cur;
	}
} //lastMatch

function quote(str) {
	return '"' + str + '"';
} //quote

function resc(str) {
	if (str == null) return "";
	return rentities(str
		.replace(/\\/gm, "\\\\")
		.replace(/-/gm, "\\-")
		.replace(/^\./gm, "\\|.")
		.replace(/\./gm, "\\.")
		.replace(/^'/gm, "\\|'")).replace('&amp;', '&');
} //resc

function rentities(str) {
	return str.replace(/&(\w+);/gm, function (match, ent) {
		var gr = {
			bull: "[ci]",
			nbsp: '~',
			copy: "(co",
			rdquo: "(rs",
			mdash: "(em",
			reg: "(rg",
			sect: "(sc",
			ge: "(>=",
			le: "(<=",
			ne: "(!=",
			equiv: "(==",
			plusmn: "(+-"
		}[ent];
		
		if (gr) return '\\' + gr;
		else return match;
	});
} //rentities

function rparseHeader(str, options) {
	var match = /([\w_.\[\]~+=@:-]+)(?:\s*)(?:\((\d\w*)\))?(?:\s*-+\s*(.*))?/.exec(str),
		name, section, text;

	if (match) {
		name = match[1];
		section = match[2];
		text = match[3];
	}

	if (!name) name = options.name || "";
	if (!section) section = options.section || "";
	if (!text) {
		if (name || section) {
			text = "";
		} else {
			text = str;
		}
	}
	if (name && text) text = " - " + text;

	return quote(resc(name.toUpperCase())) +
		" " +
		quote(section) +
		" " +
		quote(manDate(options.date)) +
		" " +
		quote(options.ver) +
		" " +
		quote(options.manual) +
		"\n.SH " +
		quote("NAME") +
		"\n\\fB" +
		name +
		"\\fR" +
		resc(text);
} //rparseHeader

function manDate(date) {
	var stamp = parseInt(date);

	if (!isNaN(stamp) && stamp.toString().length == date.length) date = stamp;
	date = new Date(date);

	if (process.env.SOURCE_DATE_EPOCH) {
		date = new Date(process.env.SOURCE_DATE_EPOCH * 1000);
	}

	var month = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"][date.getMonth()];

	return month + " " + date.getFullYear();
} //manDate

marked.defaults.format = "roff";
marked.defaults.name = "";
marked.defaults.date = new Date;
marked.defaults.section = "";
marked.defaults.ver = "";
marked.defaults.manual = "";
marked.defaults.gfm = true;
marked.defaults.breaks = false;
marked.defaults.sanitize = true;
marked.defaults.ronn = false;

module.exports = marked;
